File: lib/rag-eval.ts
import { retrieveWorkspaceContext, RetrievedChunk } from './rag'

export interface RAGEvalMetric {
  query: string
  retrievedCount: number
  topDistance: number
  meanDistance: number
  precisionAtK: number
  reciprocalRank: number
  confidenceScore: number
  latencyMs: number
  pass: boolean
}

export interface RAGEvalReport {
  timestamp: string
  workspaceId: string
  totalQueriesEvaluated: number
  overallPrecision: number
  meanReciprocalRank: number
  averageConfidence: number
  averageLatencyMs: number
  metrics: RAGEvalMetric[]
}

/**
 * Calculates Precision@K based on exact query token overlap across top-K chunks.
 */
function calculatePrecisionAtK(chunks: RetrievedChunk[], queryTokens: string[]): number {
  if (chunks.length === 0 || queryTokens.length === 0) return 0

  let relevantChunkCount = 0
  for (const chunk of chunks) {
    const chunkLower = chunk.content.toLowerCase()
    const matchesToken = queryTokens.some((token) => chunkLower.includes(token))
    if (matchesToken) {
      relevantChunkCount++
    }
  }

  return Number((relevantChunkCount / chunks.length).toFixed(2))
}

/**
 * Calculates Reciprocal Rank (1 / rank of first relevant chunk).
 */
function calculateReciprocalRank(chunks: RetrievedChunk[], queryTokens: string[]): number {
  if (chunks.length === 0 || queryTokens.length === 0) return 0

  for (let i = 0; i < chunks.length; i++) {
    const chunkLower = chunks[i].content.toLowerCase()
    if (queryTokens.some((token) => chunkLower.includes(token))) {
      return Number((1 / (i + 1)).toFixed(2))
    }
  }

  return 0
}

/**
 * Runs automated RAG evaluation benchmarks on a set of test queries for a given workspace.
 */
export async function evaluateWorkspaceRAG(
  workspaceId: string,
  testQueries: string[],
  googleApiKey: string
): Promise<RAGEvalReport> {
  const metrics: RAGEvalMetric[] = []

  for (const query of testQueries) {
    const startTime = Date.now()
    const result = await retrieveWorkspaceContext(workspaceId, query, googleApiKey)
    const latencyMs = Date.now() - startTime

    const queryTokens = query
      .toLowerCase()
      .replace(/[^a-z0-9\s]/g, '')
      .split(/\s+/)
      .filter((t) => t.length > 3)

    const precisionAtK = calculatePrecisionAtK(result.chunks, queryTokens)
    const reciprocalRank = calculateReciprocalRank(result.chunks, queryTokens)
    const distances = result.chunks.map((c) => c.distance)
    const topDistance = distances.length > 0 ? Math.min(...distances) : 1
    const meanDistance = distances.length > 0 ? distances.reduce((a, b) => a + b, 0) / distances.length : 1

    metrics.push({
      query,
      retrievedCount: result.chunks.length,
      topDistance: Number(topDistance.toFixed(3)),
      meanDistance: Number(meanDistance.toFixed(3)),
      precisionAtK,
      reciprocalRank,
      confidenceScore: Number(result.confidence.toFixed(2)),
      latencyMs,
      pass: result.chunks.length > 0 && result.confidence >= 0.35,
    })
  }

  const totalQueries = metrics.length || 1
  const overallPrecision = Number(
    (metrics.reduce((acc, m) => acc + m.precisionAtK, 0) / totalQueries).toFixed(2)
  )
  const meanReciprocalRank = Number(
    (metrics.reduce((acc, m) => acc + m.reciprocalRank, 0) / totalQueries).toFixed(2)
  )
  const averageConfidence = Number(
    (metrics.reduce((acc, m) => acc + m.confidenceScore, 0) / totalQueries).toFixed(2)
  )
  const averageLatencyMs = Math.round(
    metrics.reduce((acc, m) => acc + m.latencyMs, 0) / totalQueries
  )

  return {
    timestamp: new Date().toISOString(),
    workspaceId,
    totalQueriesEvaluated: metrics.length,
    overallPrecision,
    meanReciprocalRank,
    averageConfidence,
    averageLatencyMs,
    metrics,
  }
}

File: backend/app/services/rag_eval.py
"""Port of lib/rag-eval.ts — Precision@K / Reciprocal Rank RAG benchmark harness."""
import re
import time
from datetime import datetime, timezone

import asyncpg

from app.services import rag_service

MIN_CONFIDENCE_SCORE = 0.35


def _tokenize_query(query: str) -> list[str]:
    cleaned = re.sub(r"[^a-z0-9\s]", "", query.lower())
    return [t for t in cleaned.split() if len(t) > 3]


def _precision_at_k(chunks: list[rag_service.RetrievedChunk], query_tokens: list[str]) -> float:
    if not chunks or not query_tokens:
        return 0.0
    relevant = sum(
        1 for c in chunks if any(token in c.content.lower() for token in query_tokens)
    )
    return round(relevant / len(chunks), 2)


def _reciprocal_rank(chunks: list[rag_service.RetrievedChunk], query_tokens: list[str]) -> float:
    if not chunks or not query_tokens:
        return 0.0
    for i, chunk in enumerate(chunks):
        if any(token in chunk.content.lower() for token in query_tokens):
            return round(1 / (i + 1), 2)
    return 0.0


async def evaluate_workspace_rag(
    pool: asyncpg.Pool, workspace_id: str, test_queries: list[str], google_api_key: str
) -> dict:
    metrics = []

    for query in test_queries:
        start = time.perf_counter()
        result = await rag_service.retrieve_workspace_context(pool, workspace_id, query, google_api_key)
        latency_ms = round((time.perf_counter() - start) * 1000)

        query_tokens = _tokenize_query(query)
        precision_at_k = _precision_at_k(result.chunks, query_tokens)
        reciprocal_rank = _reciprocal_rank(result.chunks, query_tokens)
        distances = [c.distance for c in result.chunks]
        top_distance = min(distances) if distances else 1.0
        mean_distance = sum(distances) / len(distances) if distances else 1.0

        metrics.append({
            "query": query,
            "retrievedCount": len(result.chunks),
            "topDistance": round(top_distance, 3),
            "meanDistance": round(mean_distance, 3),
            "precisionAtK": precision_at_k,
            "reciprocalRank": reciprocal_rank,
            "confidenceScore": round(result.confidence, 2),
            "latencyMs": latency_ms,
            "pass": len(result.chunks) > 0 and result.confidence >= MIN_CONFIDENCE_SCORE,
        })

    total = len(metrics) or 1
    overall_precision = round(sum(m["precisionAtK"] for m in metrics) / total, 2)
    mean_reciprocal_rank = round(sum(m["reciprocalRank"] for m in metrics) / total, 2)
    average_confidence = round(sum(m["confidenceScore"] for m in metrics) / total, 2)
    average_latency_ms = round(sum(m["latencyMs"] for m in metrics) / total)

    return {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "workspaceId": workspace_id,
        "totalQueriesEvaluated": len(metrics),
        "overallPrecision": overall_precision,
        "meanReciprocalRank": mean_reciprocal_rank,
        "averageConfidence": average_confidence,
        "averageLatencyMs": average_latency_ms,
        "metrics": metrics,
    }
